### DISCUSSION TITLE
Add recipe as ingredient

### DISCUSSION THREAD

**Discussion by jtambasco** [mealie-recipes/mealie#1761] [Feature Request]
**Before submitting this feature request I have**
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
Recipes may be used as the basis to many other recipes.  For example, a pie crust recipe may be used in any pie/tart/quiche recipe.  A lasagna sheet recipe may be used for vegetarian lasagne, bolognese lasagna etc.  A vegetable stock recipe could be used in countless recipes.

**(Optional): Suggest A Solution**
* Have an option to mark a recipe as an ingredient.
* The title of that recipe is then added to the ingredient list.
* When that ingredient is added to other recipes, the ingredient list features a hyperlink to that recipe.

**Comment by michael-genson** [comment]
While not exactly what you're asking, you can add hyperlinks to your recipes that link to other recipes. That way you could do something like "Step 1: Make this recipe first" and link it.

**Reply by jtambasco** [reply]
Yeah, I had thought of doing it that way, though it feels like a bit of a workaround.  Would be awesome if it could be directly integrated!

**Reply by NicholasConrad** [reply]
I am currently using this method for my nested recipes. Make sure you always link to the publicly accessible address of the linked recipe, otherwise if you share the main recipe with someone, they won't have access to the nested one.

Another issue with this method, both recipe steps and ingredients will trigger an action (mark complete) when clicking a link in the body text of the element. Would be very convenient if it could open links in a new window, and not mark the item complete when links are clicked.

**Comment by xF4m3** [comment]
Hey, not sure if this would classify as a seperate request, but I think its more an addition to this request.

So marking an recipe as an ingredient, then adding it to other recipes as described above, should then also result in the ingredients of the sub recipe (which is marked as an ingredient) beeing added to the shopping list when the Main recipe is added. 
Not sure if that was also what @jtambasco intended but to me it sounds like this was not included.

**Reply by luixal** [reply]
IMHO, this is totally crucial for this FR. As of the next comment, directly related to #4111

**Comment by jrucker2004** [comment]
A good example of where this kind of feature makes sense:  https://www.allrecipes.com/recipe/19010/eclairs-ii/

When importing this recipe (which is actually 3 recipes: pastry, filling, and icing) the current functionality just lumps everything together.  The recipe has the ingredients for the three items separated, even though the directions aren't.

I'm not sure how this would fit into the original feature request, maybe having the ability to break up an imported recipe into multiple recipes, where the user would have to select what went where, or break up the directions manually?

In this example recipe, would that result in 4 recipes?  One for each of the three listed above, and the 4th is just "assemble these three recipes into an eclair?"  That seems like kind of a clumsy implementation, maybe the 4th recipe would display the other 3 recipes "embedded" in a pretty way that made sense, instead of just listing it as an ingredient?

**Reply by hay-kot** [reply]
This seems like a case for sections more than it would be for linked recipes

<img width="418" alt="CleanShot 2023-01-11 at 12 20 55@2x" src="https://user-images.githubusercontent.com/64056131/211920242-d120ef1c-23b1-48e3-975f-ac2d6c000a60.png">

**Reply by Wetzel402** [reply]
Do sections exist?  I haven't noticed it documented if that's the case. Thanks!

**Reply by michael-genson** [reply]
Yup! While editing a recipe, click the three dots next to the ingredient:

![image](https://github.com/user-attachments/assets/4412419c-5a23-4b5e-b327-98b0d1fa7cb2)

You can do the same for steps:

![image](https://github.com/user-attachments/assets/b2b464ce-9d35-4eec-b9f1-c620580e6420)

**Comment by tekjoey** [comment]
+1 for this feature. Just recently ran into a similar situation where I added a pizza recipe where the blogger originally linked to their pizza dough recipe. It would be great to have a generic dough recipe that could then be reused for different pizzas, calzones, etc. Additionally, I think sauces would benefit greatly from this feature. You'd have a recipe for Country Gravy, for example, that you can reference by itself, but also link to recipes for Biscuits and Gravy, Chicken Fried Steak, etc. 

While it would be cool for the importer to auto-detect these recipe "dependencies", I'd be totally ok with manually importing both and linking them. It's not a frequent need, but boy it would be cool to have.

**Comment by Inervo** [comment]
+1 for this feature as well. This would be so helpful for recipe that use the same base.
I come from the paprika app that has a "simple" implementation of this: in the ingredients, there's a line that's a link to a related recipe, and that's really usefull to not have X time the same base in each recipe.

**Comment by jtaylor54** [comment]
+1 for this feature too. I think it's a smart way to have recipes link to common rubs, sauces, fillings etc. without having to retype everything. Even a quick way to add a hyperlink to an existing recipe would be helpful.

**Comment by noahingwersen** [comment]
+1 for this feature, especially for auto adding ingredients from a sub-recipe to a shopping list. An OK workaround is adding the ingredients from the sub-recipe as a separate section in the larger recipe, then linking to the sub-recipe as a step. The main issue is if I update the sub-recipe, or want to change it to a new one. Having the ingredients linked would help a bunch with these cases

**Comment by Majawat** [comment]
I also like this, but it would be very neat to have sub-assemblies inside recipes.

For example: 
Step one: Mix together cream cheese, sour cream, mayo, cumin, chili powder. [call this now "cream cheese mixture"]
Step two: Melt butter, add corn kernels, cook for 7-8 minutes.
Step three: Add corn to [cream cheese mixture] and mix well.

This may slightly improve readability when linking ingredients

**Comment by DominikRoB** [comment]
Another +1 for this feature. Pizza (Base dough + topping) is a great example. Another usage for me would be designing bentos - some combinations of side dishes simply work better than others - and I would like to record them. 

Another usage of this, that I might use regularly, is to record 3-course dinners.. One entree recipe, one main recipe and then the dessert. Same as with the bentos: It's nice to record awesome combinations to recreate it someday.

**Reply by luixal** [reply]
In addition to pizzas (great example!): stuffed crepes, burritos, chillies (people can get really crazy here!)...

This is such a powerful addition...

**Comment by ekcom** [comment]
Even if not added as a native ingredient, being able to link to a recipe would be useful. Using this over a full hardcoded link could allow tracking through recipe renames or origin website changes.

**Comment by luixal** [comment]
Was just looking for this and I'm really surprised this hasn't been added.

May main needs would be:

 - Relate recipes somehow ( as an ingredient would be a good fit, but you would need sections from #4114 )
 - Get all ingredients added to the shopping list

:)

**Comment by FuzzyWereBeast** [comment]
This is a wonderful feature to have.  I'm surprised it hasn't already been implemented.

I have been using Tandoor.  They give you the ability to link to a recipe within the food ingredient itself.

![Screenshot 2024-10-26 081926](https://github.com/user-attachments/assets/7ba84cce-1a18-4a0f-9b74-8b06e99e18f5)

This gives you the ability to set it up once and then every recipe that uses that ingredient is automatically linked.  This is especially good for spice mixes and sauces.

![Screenshot 2024-10-26 081820](https://github.com/user-attachments/assets/aabd8a47-e1bb-4487-a000-80751c6b7518)

If/when implemented, it would be nice to be able to scale the sub recipe separately from the main recipe when adding to the shopping list.

**Comment by digital-grease** [comment]
+1 for this feature, this would be incredibly useful

**Comment by ChipSkylark37** [comment]
+1

I currently just link the recipe in steps to "make this first" but a sub-recipe or "child" would be helpful. Also taking into consideration the additional time for that recipe added to the "parent" so you get the full picture of steps saying making dough, sauce, pizza. Or having an option to skip the dough and sauce if it's already made. 

That opens up a potential new feature of "time left" so when you check off steps you can estimate how much longer is left, if each step had its own time associated with it. 

Though I haven't done a search for that I see if it's a feature request. I know that would be super helpful for pies and doughs where multiple steps of proofing and cooling may be required. Often the recipes don't include that time, but if I'm making something in the morning and didn't realize one step is "wait 3 hours", I want to realize that.

**Comment by BradFORD150** [comment]
Anything in the roadmap for adding this feature?
In addition to what others have said about auto-adding the sub-recipe ingredients to shopping lists, hyperlink to sub-recipes, accurate prep times, the other benefit would be in meal planning. If ‘pulled pork sandwich’ is added to a dinner on the meal plan, I shouldn’t have to add ‘brioche bun’ and ‘bbq sauce’ and ‘smoked pork butt’ recipes to the same meal. It can get confusing on what is an entree, what is a side, and what is a sub-recipe required to make the entree.

**Comment by parumpum** [comment]
I've started some work on this. Currently it is setup with a recipe reference on the recipes_ingredients table. This allows a recipe to exist as an ingredient for another recipe, with its own quantity.

Add button is using a modified search dialog. Shopping lists will expand sub-recipes to include all actual ingredients.

https://github.com/parumpum/mealie/tree/feat/recipe-links

**Comment by barbax7** [comment]
Also, it is possible to create sections in recipes. It could be interesting if adding recipe as ingredient will add automatically a section with all steps and all ingredients in ingredients list

**Comment by Wetzel402** [comment]
Was this feature implemented now?  I thought I saw something about it somewhere but I could be wrong.

**Reply by beanaroo** [reply]
It appears there is an open PR: https://github.com/mealie-recipes/mealie/pull/4800 (fingers crossed)

**Reply by blobberbun** [reply]
Looks like the PR failed an automatic test, was never merged, and has fallen behind…

**Reply by burgerga** [reply]
They had to wait for the ui overhaul to be merged, but now that it has, there have been recent commits, the PR author is still actively working on it ;)

**Reply by michael-genson** [reply]
Yeah we closed all PRs that haven't been worked on in a while, any open PRs will be reviewed at some point! I just got back from vacation a couple days ago so I'm catching up on real life

**Comment by michael-genson** [comment]
https://github.com/mealie-recipes/mealie/pull/4800
